File: packages/sdk/chio-ts/src/invariants/json.ts
export type JsonPrimitive = null | boolean | number | string;
export type JsonValue = JsonPrimitive | JsonValue[] | { [key: string]: JsonValue };

import { ChioInvariantError, parseJsonText } from "./errors.ts";

function compareUtf16(a: string, b: string): number {
  if (a < b) {
    return -1;
  }
  if (a > b) {
    return 1;
  }
  return 0;
}

export function canonicalizeJson(value: unknown): string {
  if (value === null) {
    return "null";
  }

  switch (typeof value) {
    case "boolean":
      return value ? "true" : "false";
    case "number":
      if (!Number.isFinite(value)) {
        throw new ChioInvariantError("canonical_json", "canonical JSON does not support non-finite numbers");
      }
      return JSON.stringify(value);
    case "string":
      return JSON.stringify(value);
    case "object":
      if (Array.isArray(value)) {
        return `[${value.map((item) => canonicalizeJson(item)).join(",")}]`;
      }

      return `{${Object.entries(value as Record<string, unknown>)
        .sort(([left], [right]) => compareUtf16(left, right))
        .map(([key, entryValue]) => `${JSON.stringify(key)}:${canonicalizeJson(entryValue)}`)
        .join(",")}}`;
    default:
      throw new ChioInvariantError(
        "canonical_json",
        `canonical JSON does not support values of type ${typeof value}`,
      );
  }
}

export function canonicalizeJsonString(input: string): string {
  return canonicalizeJson(parseJsonText(input));
}

File: packages/sdk/chio-py/src/chio/invariants/json.py
from __future__ import annotations

import json
import math
import re
from typing import Any

from ..errors import ChioInvariantError, parse_json_text

_EXPONENT_RE = re.compile(r"e([+-])0+(\d+)$")


def _canonicalize_float(value: float) -> str:
    if not math.isfinite(value):
        raise ChioInvariantError(
            "canonical_json",
            "canonical JSON does not support non-finite numbers",
        )
    if value == 0.0:
        return "0"
    rendered = format(value, ".15g").lower()
    return _EXPONENT_RE.sub(lambda match: f"e{match.group(1)}{match.group(2)}", rendered)


def canonicalize_json(value: Any) -> str:
    if value is None:
        return "null"
    if isinstance(value, bool):
        return "true" if value else "false"
    if isinstance(value, int):
        return str(value)
    if isinstance(value, float):
        return _canonicalize_float(value)
    if isinstance(value, str):
        return json.dumps(value, ensure_ascii=False, separators=(",", ":"))
    if isinstance(value, list):
        return "[" + ",".join(canonicalize_json(item) for item in value) + "]"
    if isinstance(value, dict):
        if not all(isinstance(key, str) for key in value):
            raise ChioInvariantError("canonical_json", "canonical JSON object keys must be strings")
        items = sorted(value.items(), key=lambda item: item[0].encode("utf-16-be"))
        return "{" + ",".join(
            f"{json.dumps(key, ensure_ascii=False, separators=(',', ':'))}:{canonicalize_json(entry_value)}"
            for key, entry_value in items
        ) + "}"
    raise ChioInvariantError(
        "canonical_json",
        f"canonical JSON does not support values of type {type(value).__name__}",
    )


def canonicalize_json_string(input_text: str) -> str:
    return canonicalize_json(parse_json_text(input_text))
